mypy/plugins/ctypes.py

```diff
@@ -0,0 +1,186 @@
+    """
+
+    # Every type can be converted from itself (obviously).
+    allowed_types = [tp]
+    if isinstance(tp, Instance):
+        unboxed = _find_simplecdata_base_arg(tp)
+        if unboxed is not None:
+            # If _SimpleCData appears in tp's (direct or indirect) bases, its type argument
+            # specifies the type's "unboxed" version, which can always be converted back to
+            # the original "boxed" type.
+            allowed_types.append(unboxed)
+
+            if tp.type.has_base('ctypes._PointerLike'):
+                # Pointer-like _SimpleCData subclasses can also be converted from an int or None.
+                allowed_types.append(api.named_generic_type('builtins.int', []))
+                allowed_types.append(api.named_generic_type('builtins.NoneType', []))
+
+    return UnionType.make_simplified_union(allowed_types)
+
+
+def _autounboxed_cdata(tp: Type) -> Type:
+    """Get the auto-unboxed version of a CData type, if applicable.
+
+    For *direct* _SimpleCData subclasses, the only type argument of _SimpleCData in the bases list
+    is returned.
+    For all other CData types, including indirect _SimpleCData subclasses, tp is returned as-is.
+    """
+
+    if isinstance(tp, Instance):
+        for base in tp.type.bases:
+            if base.type.fullname() == 'ctypes._SimpleCData':
+                # If tp has _SimpleCData as a direct base class,
+                # the auto-unboxed type is the single type argument of the _SimpleCData type.
+                assert len(base.args) == 1
+                return base.args[0]
+    # If tp is not a concrete type, or if there is no _SimpleCData in the bases,
+    # the type is not auto-unboxed.
+    return tp
+
+
+def _get_array_element_type(tp: Type) -> Optional[Type]:
+    """Get the element type of the Array type tp, or None if not specified."""
+
+    if isinstance(tp, Instance):
+        assert tp.type.fullname() == 'ctypes.Array'
+        if len(tp.args) == 1:
+            return tp.args[0]
+    return None
+
+
+def array_constructor_callback(ctx: 'mypy.plugin.FunctionContext') -> Type:
+    """Callback to provide an accurate signature for the ctypes.Array constructor."""
+
+    # Extract the element type from the constructor's return type, i. e. the type of the array
+    # being constructed.
+    et = _get_array_element_type(ctx.default_return_type)
+    if et is not None:
+        allowed = _autoconvertible_to_cdata(et, ctx.api)
+        assert len(ctx.arg_types) == 1, (
+            "The stub of the ctypes.Array constructor should have a single vararg parameter")
+        for arg_num, arg_type in enumerate(ctx.arg_types[0], 1):
+            if not is_subtype(arg_type, allowed):
+                ctx.api.msg.fail(
+                    'Array constructor argument {} of type "{}"'
+                    ' is not convertible to the array element type "{}"'
+                    .format(arg_num, arg_type, et),
+                    ctx.context)
```

Comment:
You're right, it doesn't work with `*args` at the moment, but it definitely should. `**kwargs` should always be an error, since `Array` doesn't take keyword arguments. (Yes, `**{}` is technically allowed, but I *really* hope nobody does that.)

I'm not sure how to handle this in a function hook though - I don't see a way to find out if an argument is a regular arg, `*args`, or `**kwargs`.

In any case, I've added a (currently failing) test for `*args`.

Reply:
Yes, the logic in function hook is sub-optimal, this is probably the most problematic hook currently, several people who tried it, found the API inconvenient. You also get corresponding expressions in `ctx.args[0]` for every type in `ctx.arg_types[0]`, maybe you can try checking if it is a `StarExpr(...)` and then somehow "unpack" type if it is maps to iterable supertype?

Reply:
> maybe you can try checking if it is a `StarExpr(...)`

This doesn't seem to work - if I print out `ctx.args[0][0]` for a call like `intarr4(*int_values)`, I get `NameExpr(int_values [__main__.int_values])`, with no `StarExpr` around it.

Reply:
Hm... This is unfortunate, there is a trivial fix to `apply_function_plugin()` to pass `arg_kinds` to `FunctionContext`, but this looks like a partial solution of a bigger problem. Could you please add a `TODO` comment here? We can fix it later after we decide what to do with the `FunctionContext` API.